strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,957 @@
+        deltas = self.means[:, mode_ind] - alpha_mean
+        # Construct sum of coherent state covariance matrix and all covariances in the state
+        cov_sum = (
+            self.covs[:, mode_ind, :][:, :, mode_ind] + self.hbar * np.eye((len(mode_ind))) / 2
+        )
+        # Sum all Gaussian peaks
+        exp_arg = np.einsum("...j,...jk,...k", deltas, np.linalg.inv(cov_sum), deltas)
+        weighted_exp = (
+            np.array(self.weights)
+            * self.hbar ** len(modes)
+            * np.exp(-0.5 * exp_arg)
+            / np.sqrt(np.linalg.det(cov_sum))
+        )
+        fidelity = np.sum(weighted_exp)
+        return fidelity
+
+    def fidelity_vacuum(self, modes=None):
+        r"""Returns the fidelity to the vacuum.
+
+        Args:
+            modes (list or None): modes to use for fidelity calculation
+
+        Returns:
+            float: fidelity of the state in modes to the vacuum
+        """
+        if modes is None:
+            modes = self.get_modes()
+        alpha = np.zeros(len(modes))
+        fidelity = self.fidelity_coherent(alpha, modes=modes)
+        return fidelity
+
+    def parity_val(self, modes=None):
+        r"""Returns the expectation value of the parity operator, which is the
+        value of the Wigner function at the origin.
+
+        Args:
+            modes (list or None): modes to use for parity calculation
+
+        Returns:
+            float: parity of the state in modes
+        """
+        if modes is None:
+            modes = list(range(self.nlen))
+        # Sort by (q1,p1,q2,p2,...)
+        mode_ind = np.sort(np.append(2 * np.array(modes), 2 * np.array(modes) + 1))
+        exp_arg = np.einsum(
+            "...j,...jk,...k",
+            self.means[:, mode_ind],
+            np.linalg.inv(self.covs[:, mode_ind, :][:, :, mode_ind]),
+            self.means[:, mode_ind],
+        )
+        weighted_exp = (
+            np.array(self.weights)
+            * np.exp(-0.5 * exp_arg)
+            / np.sqrt(np.linalg.det(self.covs[:, mode_ind, :][:, :, mode_ind]))
+        )
+        parity = np.sum(weighted_exp) * (self.hbar / 2) ** len(modes)
+        return parity
+
+    def loss(self, T, k):
+        r"""Implements a loss channel in mode k. T is the loss parameter that must be
+        between 0 and 1.
+
+        Args:
+            T (float): loss amount is \sqrt{T}
+            k (int): mode that loses energy
+
+        Raises:
+            ValueError: if the mode is not in the list of active modes
+        """
+
+        if self.active[k] is None:
+            raise ValueError("Cannot apply loss channel, mode does not exist")
+
+        X = np.sqrt(T) * np.identity(2)
+        Y = self.hbar * (1 - T) * np.identity(2) / 2
+        X2, Y2 = self.expandXY([k], X, Y)
+        self.apply_channel(X2, Y2)
+
+    def thermal_loss(self, T, nbar, k):
+        r"""Implements the thermal loss channel in mode k. T is the loss parameter that must
+        be between 0 and 1.
```

Comment:
Should it be checked that `T` is within the correct range?

Reply:
Will do.